src/action.py

```diff
@@ -0,0 +1,344 @@
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Module for taking the required actions to match the server state with the local state."""
+
+import logging
+import typing
+
+from . import exceptions, reconcile, types_
+from .discourse import Discourse
+
+DRAFT_NAVLINK_LINK = "<not created due to draft mode>"
+FAIL_NAVLINK_LINK = "<not created due to error>"
+DRAFT_MODE_REASON = "draft mode"
+NOT_DELETE_REASON = "delete_topics is false"
+
+
+def _create(
```

Comment:
This is a preference, i.e., I think reasonable developers may disagree on this (depending on their preference of functional versus object oriented programming). The defined types here are really just supposed to be simple data structures to hold data rather than include business logic (I know some of the have simple functions in them, so I do bend preferences where I think it makes a meaningful difference in testability, readability etc). I prefer to write business logic in separate functions because of the easy input/ output testing that can be applied to them. Whilst, in this case, it means that there are these routing functions that are needed, the difference between the approaches is relatively minor. Switching to it at this point would also require quite a bit of changes

Reply:
Yeap I agree that's it's a matter of preference, coming from the Java world I tend to design things OO style. In my opinion the best approach would be to have interfaces that defines the class data structure and implementation of the interface with the `execute` method.
I don't know how common it is to use interfaces or abstract classes in Python

Reply:
It is possible although I don't see it done a lot. Generally speaking, Python developers prefer functions on modules and do OOP when they have to based on what I hear listening to podcasts, at conferences and reading code. Maybe because Python is prevalent in web development where the popular frameworks encourage this style (my background is also from that space although I have developed a lot of non web code as well.) Closing this for now, please let me know if there is further feedback.